include/swift/AST/Requirement.h

```diff
@@ -124,6 +124,55 @@ struct StructuralRequirement {
   bool inferred = false;
 };
 
+/// Represents an invalid requirement, such as `T: Int`.
+///
+/// Invalid requirements are recorded while computing the
+/// generic signature of a declaration, and diagnosed via
+/// \c diagnoseRequirementErrors .
+struct RequirementError {
+  /// The kind of requirement error.
+  enum class Kind {
+    InvalidConformance,
+    ConcreteTypeMismatch,
+    NonTypeParameter,
+    RedundantRequirement,
+  } kind;
+
+  /// The invalid requirement.
+  Requirement requirement;
+
+  SourceLoc loc;
+
+private:
+  RequirementError(Kind kind, Requirement requirement, SourceLoc loc)
+    : kind(kind), requirement(requirement), loc(loc) {}
+
+public:
+  static RequirementError forInvalidConformance(Type subjectType,
```

Comment:
Perhaps forInvalidTypeRequirement() is a better name -- in this case it's not a conformance *or* a superclass requirement, it's just something like 'T : Int' right?